src/algebra/big_operators/order.lean

```diff
@@ -308,3 +309,50 @@ open opposite
 (@unop_add_hom β _).map_sum _ _
 
 end with_top
+
+namespace finset
+
+/--
+The strong pigeonhole principle for finitely many pigeons and pigeonholes.
+-/
+lemma strong_pigeonhole {s : finset α} {t : finset β} (hne : t.nonempty) [decidable_eq β]
+  (f : α → β) (hf : ∀ a ∈ s, f a ∈ t)
+  (n : ℕ) (hn : t.card * n ≤ s.card) :
+  ∃ y ∈ t, n ≤ (s.filter (λ x, f x = y)).card :=
+begin
+  classical, by_contra hz, push_neg at hz,
+  suffices h : s.card < t.card * n,
+  have h' := lt_of_lt_of_le h hn,
+  exact nat.lt_asymm h' h',
+
+  have key : ∀ y ∈ t, (s.filter (λ x, f x = y)).card ≠ 0 → y ∈ s.image f,
```

Comment:
This seems a generally useful fact, so should be proved separately.

Reply:
:+1: added `finset.preimage_card_ne_zero_iff_mem_image`